tests/integration/tls/helpers_manual_tls.py

```diff
@@ -0,0 +1,209 @@
+
+        """
+        logging.info("Getting outstanding certificate requests")
+        action = await self.tls_unit.run_action("get-outstanding-certificate-requests")
+        action: Action = await action.wait()
+        if action.status != "completed":
+            message = action.safe_data.get(
+                "message",
+                "Failed to get outstanding certificate requests",
+            )
+            if "No outstanding certificate requests" in message:
+                logging.info(message)
+            if "No certificates relation has been created yet" in message:
+                logging.info(message)
+            logging.error(message)
+            raise GettingOutstandingCertificateRequestsFailedError(message)
+        csrs = json.loads(action.results["result"])
+        self.csr_queue = deque([CSR.from_charm_csr(csr) for csr in csrs])
+
+    @retry(
+        wait=wait_exponential(multiplier=1, min=5, max=20),
+        stop=stop_after_attempt(100),
+    )
+    async def wait_for_csrs_in_queue(self, nbr_csrs: int = 1) -> None:
+        """Wait for the number of csrs in the queue to be equal to the number of csrs specified.
+
+        Arguments:
+        ---------
+            nbr_csrs : int
+                The number of csrs to wait for in the queue.
+
+        Raises:
+        ------
+            CSRsMissingError: If the number of csrs in the queue is less than the expected number.
+
+        """
+        await self.get_outstanding_certificate_requests()
+        if len(self.csr_queue) < nbr_csrs:
+            message = f"{nbr_csrs - len(self.csr_queue)} CSRs missing in queue"
+            raise CSRsMissingError(message)
+
+    async def process_csr(self, csr: CSR) -> None:
+        """Process the certificate signing request.
+
+        Arguments:
+        ---------
+            csr : CSR
+                The certificate signing request to process.
+
+        Raises:
+        ------
+            ProvidingCertificateFailedError: If providing the certificate fails.
+
+        """
+        # Generate a certificate
+        certificate = generate_certificate(
+            csr=csr.csr,
+            ca=self.ca,
+            ca_key=self.ca_key,
+            is_ca=csr.is_ca,
+        )
+        logger.info("Generated certificate for %s", csr.unit_name)
+        # Send the certificate back to the charm
+        action = await self.tls_unit.run_action(
+            "provide-certificate",
+            relation_id=csr.relation_id,
+            **{
+                "certificate": base64.b64encode(certificate).decode(),
+                "ca-certificate": base64.b64encode(self.ca).decode(),
+                "certificate-signing-request": base64.b64encode(
+                    csr.csr,
+                ).decode(),
+            },
+        )
+        action = await action.wait()
+        if action.status != "completed":
+            message = "Failed to provide certificate for %s", csr.unit_name
```

Comment:
This creates a tuple
```suggestion
            message = f"Failed to provide certificate for {csr.unit_name}"
```